sled-agent/src/zone_bundle.rs

```diff
@@ -6,42 +6,466 @@
 
 //! Tools for collecting and inspecting service bundles for zones.
 
-use crate::params::ZoneBundleCause;
-use crate::params::ZoneBundleMetadata;
+use crate::storage_manager::StorageResources;
+use anyhow::anyhow;
 use camino::FromPathBufError;
 use camino::Utf8Path;
 use camino::Utf8PathBuf;
+use chrono::DateTime;
+use chrono::Utc;
 use flate2::bufread::GzDecoder;
 use illumos_utils::running_zone::is_oxide_smf_log_file;
 use illumos_utils::running_zone::RunningZone;
+use illumos_utils::zfs::ZFS;
 use illumos_utils::zone::AdmError;
+use schemars::JsonSchema;
+use serde::Deserialize;
+use serde::Serialize;
 use slog::Logger;
+use std::cmp::Ord;
+use std::cmp::Ordering;
+use std::cmp::PartialOrd;
+use std::collections::BTreeMap;
+use std::collections::BTreeSet;
+use std::collections::HashSet;
 use std::io::Cursor;
+use std::path::Path;
 use std::path::PathBuf;
+use std::sync::Arc;
+use std::time::Duration;
 use std::time::SystemTime;
 use tar::Archive;
 use tar::Builder;
 use tar::Header;
+use tokio::process::Command;
+use tokio::sync::Mutex;
+use tokio::sync::Notify;
+use tokio::time::sleep;
+use tokio::time::Instant;
 use uuid::Uuid;
 
-/// Context for creating a bundle of a specified zone.
-#[derive(Debug, Default)]
-pub struct ZoneBundleContext {
-    /// The directories into which the zone bundles are written.
-    pub storage_dirs: Vec<Utf8PathBuf>,
-    /// The reason or cause for creating a zone bundle.
+/// An identifier for a zone bundle.
+#[derive(
+    Clone,
+    Debug,
+    Deserialize,
+    Eq,
+    Hash,
+    JsonSchema,
+    Ord,
+    PartialEq,
+    PartialOrd,
+    Serialize,
+)]
+pub struct ZoneBundleId {
+    /// The name of the zone this bundle is derived from.
+    pub zone_name: String,
+    /// The ID for this bundle itself.
+    pub bundle_id: Uuid,
+}
+
+/// The reason or cause for a zone bundle, i.e., why it was created.
+#[derive(
+    Clone,
+    Copy,
+    Debug,
+    Default,
+    Deserialize,
+    Eq,
+    Hash,
+    JsonSchema,
+    PartialEq,
+    Serialize,
+)]
+#[serde(rename_all = "snake_case")]
+#[non_exhaustive]
+pub enum ZoneBundleCause {
+    /// Some other, unspecified reason.
+    #[default]
+    Other,
+    /// A zone bundle taken when a sled agent finds a zone that it does not
+    /// expect to be running.
+    UnexpectedZone,
+    /// An instance zone was terminated.
+    TerminatedInstance,
+    /// Generated in response to an explicit request to the sled agent.
+    ExplicitRequest,
+}
+
+impl PartialOrd for ZoneBundleCause {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for ZoneBundleCause {
+    fn cmp(&self, other: &Self) -> Ordering {
+        use ZoneBundleCause::*;
+        match (self, other) {
```

Comment:
That's what I had originally, but thought it was more clear to have them explicit. But it's definitely more code, so I'm ambivalent. Deriving is fine, as long as I drop a note that the variants shouldn't be reordered without consideration.

Reply:
I think a comment that the order actually matters + a unit test that would fail if someone ignored the comment (which you already have!) would be enough guard rails to use `derive`. In general I'm with you that explicit is better than implicit, but in this case the nested match is hard enough to read that it actually seems _less_ clear than a comment saying "these are ordered in source code order and that actually matters", IMO.